Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift

```diff
@@ -291,6 +291,70 @@ public final class SwiftTargetBuildDescription {
         }
 
         try self.generateResourceEmbeddingCode()
+        try self.generateTestObservation()
+    }
+
+    private func generateTestObservation() throws {
+        guard target.type == .test, buildParameters.targetTriple.isDarwin(), buildParameters.experimentalTestOutput else {
+            return
+        }
+
+        // FIXME: Ensure we don't make multiple of these.
+
+        let content =
+            """
+            import Foundation
+            import XCTest
+
+            public final class SwiftPMXCTestObserver: NSObject {
+                public override init() {
+                    super.init()
+                    XCTestObservationCenter.shared.addTestObserver(self)
+                }
+            }
+
+            struct FailureRecord: Encodable {
+                let location: Location?
+                let testCase: String
+                let message: String
+            }
+
+            struct Location: Encodable {
+                let file: String
+                let line: Int
+            }
+
+            extension SwiftPMXCTestObserver: XCTestObservation {
+                public func testCase(_ testCase: XCTestCase, didFailWithDescription description: String, inFile: String?, atLine: Int) {
+                    let location: Location?
+                    if let file = inFile {
+                        location = Location(file: file, line: atLine)
+                    } else {
+                        location = nil
+                    }
+
+                    let outputPath = "\(buildParameters.testOutputPath)"
+                    let record = FailureRecord(location: location, testCase: testCase.description, message: description)
+                    if let data = try? JSONEncoder().encode(record) {
+                        if let fileHandle = FileHandle(forWritingAtPath: outputPath) {
+                            defer { fileHandle.closeFile() }
+                            fileHandle.seekToEndOfFile()
+                            fileHandle.write("\\n".data(using: .utf8)!)
+                            fileHandle.write(data)
+                        } else {
+                            _ = try? data.write(to: URL(fileURLWithPath: outputPath))
+                        }
```

Comment:
Need to have synchronization between parallel tests here.

Reply:
Same as above, seems fine in practice. Maybe we can just merge this as experimental already.

Reply:
I managed to construct an example now where the unsynchronized writes are problematic.

Reply:
Fixed now.